src/fparser/common/readfortran.py

```diff
@@ -614,12 +617,48 @@ def find_module_source_file(self, mod_name):
 
     def set_format(self, mode):
         """
-        Set Fortran code mode (fixed/free format etc).
+        Set Fortran code mode (fixed/free format etc). If handling of
+        OMP sentinels is also enabled, this function will also create
+        the required regular expressions to handle conditional sentinels
+        depending on the (new) format
 
         :param mode: Object describing the desired mode for the reader
         :type mode: :py:class:`fparser.common.sourceinfo.FortranFormat`
         """
         self._format = mode
+        if not self._omp_sentinel:
+            return
+
+        if self._format.is_fixed or self._format.is_f77:
+            sentinel = r"^([\!\*c]\$)"
+            # Initial lines fixed format sentinels: !$, c$, *! in first
+            # column then only spaces and digits up to column 5, and a
+            # space or 0 at column 6
+            init_line = r"[ 0-9]{3}[ 0]"
+            # Continued lines fixed format sentinels: !$, c$, *! in first
+            # columns, then three spaces, and a non-space non-0 character
+            # in column 6:
+            cont_line = r"   [^ 0]"
+            # Combine these two regular expressions
+            self._re_omp_sentinel = re.compile(
+                f"{sentinel}({init_line}|{cont_line})", re.IGNORECASE
+            )
+        else:
+            # Initial free format sentinels: !$ as the first non-space
+            # character followed by a space.
+            self._re_omp_sentinel = re.compile(r"^ *(\!\$) ", re.IGNORECASE)
+            # Continued lines free format sentinels: !$ as the first non-space
+            # character with an optional & that can have spaces or not. The
+            # important implication of the latter is that a continuation line
+            # can only be properly detected if the previous line had a
+            # sentinel (since it does not require a space after the sentinel
+            # anymore. Without the requirement of a space, the regular
+            # expression for continuation lines will also match !$omp
```

Comment:
Difficult question - I believe according to the standard we can't really exclude this, though I am not sure if the standard is that precise tbh:
Free source (normal omp) directive (3.1.2):

> Initial directive lines must have a space after the sentinel ... Thus, continued directive lines must have an ampersand (&) as the last non-blank character on the line, prior to any comment placed inside the directive; continuation directive lines can have an ampersand after the directive sentinel with optional white space before and after the ampersand.

Both ifort and gfortran therefore accept:
```
!$omp parallel &
!$ompdo
```
So we would have to exclude anything starting with `!$omp` (without space after omp!!!).

About conditional sentinels:
> Initial lines must have a blank character after the sentinel; and Continued lines must have an ampersand as the last non-blank character on the line[...] Continuation lines can have an ampersand after the sentinel, with optional white space before and after the ampersand. If these criteria are met, the sentinel is replaced by two spaces. 

So, imho this should mean that:
```
!$ integer, parameter :: &
!$omp = 2
```
is a valid conditional (since white space is optional before and after the ampersand ... though I think it should be specified explicitly how white space without & are handled, but I would assume if the standard does not specify that there must be a white space, it is not required).

Now the truth is, both ifort and gfortran reject the above example. But both examples:
```
!$ integer :: &
!$ompvar

!$ character(7) :: xx = "my &
!$omp "
```
is accepted by both gfortran and intel, so I think I can't exclude "!$omp".